dialogue-clients/src/main/java/com/palantir/dialogue/clients/DefaultDialogueDnsResolver.java

```diff
@@ -41,8 +45,72 @@ public ImmutableSet<InetAddress> resolve(String hostname) {
             }
             return ImmutableSet.copyOf(results);
         } catch (UnknownHostException e) {
-            log.warn("Unknown host '{}'", UnsafeArg.of("hostname", hostname), e);
+            ExtractedGaiError gaiError = extractGaiErrorString(e);
+            log.warn(
+                    "Unknown host '{}'",
+                    SafeArg.of("gaiErrorMessage", gaiError.getErrorMessage()),
+                    SafeArg.of("gaiErrorType", gaiError.getErrorType()),
+                    UnsafeArg.of("hostname", hostname),
+                    e);
             return ImmutableSet.of();
         }
     }
+
+    // these strings were taken from glibc-2.39, but likely have not changed in quite a while
+    // strings may be different on BSD systems like macos
+    // TODO(dns): update this list to try to match against known strings on other platforms
+    private static final Map<String, String> EXPECTED_GAI_ERROR_STRINGS = ImmutableMap.<String, String>builder()
+            .put("Address family for hostname not supported", "EAI_ADDRFAMILY")
+            .put("Temporary failure in name resolution", "EAI_AGAIN")
+            .put("Bad value for ai_flags", "EAI_BADFLAGS")
+            .put("Non-recoverable failure in name resolution", "EAI_FAIL")
+            .put("ai_family not supported", "EAI_FAMILY")
+            .put("Memory allocation failure", "EAI_MEMORY")
+            .put("No address associated with hostname", "EAI_NODATA")
+            .put("Name or service not known", "EAI_NONAME")
+            .put("Servname not supported for ai_socktype", "EAI_SERVICE")
+            .put("ai_socktype not supported", "EAI_SOCKTYPE")
+            .put("System error", "EAI_SYSTEM")
+            .put("Processing request in progress", "EAI_INPROGRESS")
+            .put("Request canceled", "EAI_CANCELED")
+            .put("Request not canceled", "EAI_NOTCANCELED")
+            .put("All requests done", "EAI_ALLDONE")
+            .put("Interrupted by a signal", "EAI_INTR")
+            .put("Parameter string not correctly encoded", "EAI_IDN_ENCODE")
+            .put("Result too large for supplied buffer", "EAI_OVERFLOW")
+            .buildOrThrow();
+
+    @Value.Immutable
+    interface ExtractedGaiError {
+        @Value.Parameter
+        String getErrorMessage();
+
+        @Value.Parameter
+        String getErrorType();
+    }
+
+    private static ExtractedGaiError extractGaiErrorString(UnknownHostException exception) {
+        if (exception == null) {
+            return ImmutableExtractedGaiError.of("null", "null");
+        }
+
+        try {
+            StackTraceElement[] trace = exception.getStackTrace();
+            if (trace.length > 0) {
+                StackTraceElement top = trace[0];
+                if ("java.net.InetAddress$CachedLookup".equals(top.getClassName())) {
+                    return ImmutableExtractedGaiError.of("cached", "cached");
+                }
```

Comment:
I think the cached case failures can be checked most easily using `Objects.equals(requestedHostname, exception.getMessage())` -- either way we should add a test for this.